nemo/collections/llm/gpt/data/core.py

```diff
@@ -1003,33 +1042,55 @@ def _process_example(self, example):
                 self.num_turn_start_tokens,
             )
         else:
-            if "conversations" in example:  # convert ShareGPT format to HuggingFace chat template format
-                example = _transform_to_chat_message(example)
-            result = _preprocess_hf_chat_template(
-                example,
-                self.tokenizer,
-            )
+            result = _chat_preprocess(example, self.tokenizer, self.tool_schemas)
+
         # store metadata in dataset, in case user may have keys required in the prediction json files
         metadata = {k: v for k, v in example.items() if k not in ['conversations', 'messages']}
+
         result['metadata'] = metadata
         if self.output_original_text:
-            result['metadata']['conversations'] = example['conversations']
+            for k in ['conversations', 'messages']:
+                if k in example:
+                    result['metadata'][k] = example[k]
 
         return result
 
     def collate_fn(self, batch):
-        input_ids = [item['input_ids'][:-1].tolist() for item in batch]
-        labels = [item['input_ids'][1:].tolist() for item in batch]
-        loss_mask = [item['mask'][1:].tolist() for item in batch]
+        """Overrides parent function to ensure loss mask gets exported correctly."""
+        # Removes the last token from each input sequence to ensure the model
+        # never sees the token it is supposed to predict. This enforces an
+        # autoregressive training setup where the model learns to generate
+        # the next token step-by-step.
+        input_ids = [item["input_ids"][:-1] for item in batch]
+        # Removes the first token from each input sequence to create labels
+        # that align with the model's prediction target. This ensures that
+        # at time step `t`, the model's output is evaluated against the token
+        # that originally followed the input at `t` in the dataset.
+        labels = [item["input_ids"][1:] for item in batch]
+        # Context tokens remain unchanged, representing the initial portion of
+        # the sequence that serves as input to the model. This allows the model
+        # to condition its predictions on prior information.
+        contexts = [item["context_ids"] for item in batch]
+        # Extracts the assistant's response portion of the sequence, which
+        # represents the part the model is trained to generate. This helps
+        # distinguish between the input prompt and the expected model output.
+        answers = [item["answer_ids"] for item in batch]
+        # Removes the first element from the mask to align with the shifted labels,
+        # ensuring that loss is only computed for valid, predictable tokens. This
+        # prevents the model from incurring loss on tokens that were never meant to
+        # be predicted, such as user-provided context or padding.
+        loss_mask = [item["loss_mask"][1:] for item in batch]
```

Comment:
can the comments be more concise?

Reply:
The comments are helpful for someone modifying these parameters in the future. Why do they need to be more concise?